docs/IcepeakProtocolMultiSub.md

```diff
@@ -0,0 +1,177 @@
+  "type": "object",
+  "properties": {
+    "type": { "const": "subscribe" },
+    "paths": { "type": "array", "items": { "type": "string" } },
+    "token": { "type": "string" }
+  }
+}
+```
+
+### Server Subscribe Response
+The server sends back a payload to the client. The payload will always contain a status code:
+
+| Status code | When                                  |
+| ---- | -------------------------------------------- |
+| 200  | Subscription was successfully processed      |
+| 400  | Request payload was malformed                |
+| 401  | No authorization token provided              |
+| 403  | Authorization token was rejected / malformed |
+
+
+If the status code is `200` and **whether or not the client is already subscribed**, the client can expect a payload from server that contains:
+- The status code.
+- Path(s) of the subscription requested.
+- The current value(s) of that path(s).
+
+`JSON Schema` declaration of the server response:
+```javascript
+{
+  "$schema": "https://json-schema.org/draft/2020-12/schema",
+  "$id": "wss://updates.channable.com",
+  "title": "Subscription status",
+  "description": "Indicates whether the subscription was successful",
+  "type": "object",
+  "properties": {
+    "type": { "const": "subscribe" },
+    "paths": {
+      "type": "array",
+      "items": {
+        "type": "object",
+        "properties": {
+          "path": { "type": "string" },
+          "value": {}
+        }
+      }
+    },
+    "code": { "type": "number" },
+    "message": { "type": "string" },
+    "extra": {}
+  }
+}
+```
+
+## Unsubscribe
+
+In summary:
+- The client can send a payload that contains paths to unsubscribe from.
+- The client can expect a response from the server that contains the status/acknowledgement of the request.
+  - In the case of a successful request, the response also contains the list of paths in the client request.
+
+### Client Unsubscribe Request
+
+`JSON Schema` declaration of the unsubscribe client request:
+```javascript
+{
+  "$schema": "https://json-schema.org/draft/2020-12/schema",
+  "$id": "wss://updates.channable.com",
+  "title": "Remove Subscription",
+  "description": "Remove an existing subscription",
+  "type": "object",
+  "properties": {
+    "type": { "const": "unsubscribe" },
+    "paths": { "type": "array", "items": { "type": "string" } }
+  }
+}
+```
+
+### Server Unsubscribe Response
+The server sends back a payload to the client. The payload will always contain a status code:
+
+| Status code   | When                             |
+| ------------- | -------------------------------- |
+| 200  | Unsubscription was successfully processed |
+| 400  | Request payload was malformed             |
+
+If the status code is `200` and **whether or not the client is already unsubscribed**, the client can expect a payload from server that contains:
+- the list of the unsubscribe paths that the client had sent in the request.
```

Comment:
I'm not sure this will work out elegantly:

1. This would cause the requests to be less idempotent (in terms of the response received, though this in itself is up for debate if its desireable).
2. There is no such case where we unsubscribe from only a subset of paths in the request.
3. We don't have logic where we individually check if the paths exist. We just do an unsubscribe on all paths, there is no underlying success/failure indication, and it seems to work fine (according to the tests)
  - [Test](https://github.com/channable/icepeak/blob/7cc9362063417d5fe6701058a74032e6640d1a4b/server/tests/Icepeak/Server/MultiSubscriptionSpec.hs#L295-L304) 
    - Sidenote: I should definitely make the tests more granular

If we were to implement this, there would be more logic in the unsubscribes.
  - [Unsubscribe logic](https://github.com/channable/icepeak/blob/7cc9362063417d5fe6701058a74032e6640d1a4b/server/src/Icepeak/Server/WebsocketServer/MultiSubscription.hs#L170-L179)

But yes, I agree it would be more useful. Is it worth it to in introduce subscription detection logic?

Reply:
> This would cause the requests to be less idempotent 

Idempotency usually refers to the effective result of an operation. Returning different responses (as long as the effect of the request is the same) should be fine, and is more informative from a user's perspective.

> There is no such case where we unsubscribe from only a subset of paths in the request.

While that is true, it doesn't prohibit clients from making mistakes in their subscriptions, i.e. they may try to unsubscribe twice from a path from different code flows. In that case it would be nice to get some feedback on what exactly happened.

> We don't have logic where we individually check if the paths exist.

Can't we add a simple check in the following within the `modifyMVar`? We only need to know whether that specific connection was actually subscribed to the path. Gathering them and sending them off to the client should be fairly straightforward.

https://github.com/channable/icepeak/blob/7cc9362063417d5fe6701058a74032e6640d1a4b/server/src/Icepeak/Server/WebsocketServer/MultiSubscription.hs#L174-L175

Reply:
Makes sense :+1: 

Though suddenly another thing I realise might be bad about this is that the client won't have a guarantee to identify which unsubscribe request the response originated from.

Is this a concern?

Reply:
> Is this a concern?

I don't think so, if the frontend choses to depend on that behavior, they'll probably implement a connection/communication manager

> We probably do the same for subscribes?

Good idea